src/clang-cpp-frontend/clang_cpp_convert.h

```diff
@@ -148,6 +161,250 @@ class clang_cpp_convertert : public clang_c_convertert
    * need to build an new_object when dealing with constructor_call
    */
   bool need_new_object(const clang::Stmt *parentStmt);
+
+  /*
+   * Methods for virtual tables and virtual pointers
+   *  TODO: add link to wiki page
+   */
+  std::string vtable_type_prefix = "virtual_table::";
+  std::string vtable_ptr_suffix = "@vtable_pointer";
+  std::string thunk_prefix = "thunk::";
+  using function_switch = std::map<irep_idt, exprt>;
+  using switch_table = std::map<irep_idt, function_switch>;
+  /*
+   * traverse methods to:
+   *  1. convert virtual methods and add them to class' type
+   *  2. If there is no existing vtable type symbol,
+   *    add virtual table type symbol in the context
+   *    and the virtual pointer.
+   *    Then add a new entry in the vtable.
+   *  3. If vtable type symbol exists, do not add it or the vptr,
+   *    just add a new entry in the existing vtable.
+   *  4. If method X overrides a base method,
+   *    add a thunk function that does late casting of the `this` parameter
+   *    and redirects the call to the overriding function (i.e. method X itself)
+   *
+   *  Last but not least, set up the vtable varaible symbols (i.e. these are the struct variables
+   *  instantiated from the vtable type symbols)
+   */
+  bool get_struct_class_virtual_methods(
+    const clang::CXXRecordDecl *cxxrd,
+    struct_typet &type);
+  /*
+   * additional annotations for virtual or overriding methods
+   */
+  bool annotate_virtual_overriding_methods(
+    const clang::CXXMethodDecl *md,
+    struct_typet::componentt &comp);
+  /*
+   * Check the existence of virtual table type symbol.
+   * If it exists, return its pointer. Otherwise, return nullptr.
+   */
+  symbolt *check_vtable_type_symbol_existence(struct_typet &type);
```

Comment:
I'd personally recommend against prefixing predicates with `check_`. Something like `is_` or `get_` is more common.

Also, why is the `type` parameter given by non-const reference? Is it to be modified?

Reply:
> Also, why is the type parameter given by non-const reference? Is it to be modified?
yes

Reply:
> yes

Strange, the implementation
```c++
symbolt *
clang_cpp_convertert::check_vtable_type_symbol_existence(struct_typet &type)
{
  irep_idt vt_name = vtable_type_prefix + tag_prefix + type.tag().as_string();
  return context.find_symbol(vt_name);
}
```
does not seem to modify it and neither does the documentation hint to that.

Reply:
Spot on
Apologies for the confusing answer I provided earlier. It's probably a copy-and-past mistake I've made.
Now fixed. I've just added `const`.